src/main/scala/vsys/state/opcdiffs/AssertOpcDiff.scala

```diff
@@ -0,0 +1,103 @@
+      Right(OpcDiff.empty)
+    else
+      Left(GenericError(s"Invalid Assert (gteq0): Value ${Longs.fromByteArray(v.data)} is negative"))
+  }
+
+  def ltEq(v1: DataEntry, v2: DataEntry): Either[ValidationError, OpcDiff] = {
+    if (v1.dataType == DataType.Amount && v2.dataType == DataType.Amount
+      && Longs.fromByteArray(v1.data) <= Longs.fromByteArray(v2.data))
+      Right(OpcDiff.empty)
+    else
+      Left(GenericError(s"Invalid Assert (lteq0): Value ${Longs.fromByteArray(v2.data)} is larger than $v1"))
+  }
+
+  def ltInt64(m: DataEntry): Either[ValidationError, OpcDiff] = {
+    if (m.dataType == DataType.Amount && Longs.fromByteArray(m.data) <= Long.MaxValue)
+      Right(OpcDiff.empty)
+    else
+      Left(GenericError(s"Invalid Assert (ltint64): Value ${Longs.fromByteArray(m.data)} is invalid"))
+  }
+
+  def gt0(v: DataEntry): Either[ValidationError, OpcDiff] = {
+    if (v.dataType == DataType.Amount && Longs.fromByteArray(v.data) > 0)
+      Right(OpcDiff.empty)
+    else
+      Left(GenericError(s"Invalid Assert (gt0): Value $v is non-positive"))
+  }
+
+  def eq(add1: DataEntry, add2: DataEntry): Either[ValidationError, OpcDiff] = {
+    if (add1.dataType == DataType.Address && add2.dataType == DataType.Address
+      && Address.fromBytes(add1.data) == Address.fromBytes(add2.data))
+      Right(OpcDiff.empty)
+    else if (add1.dataType == DataType.Amount && add2.dataType == DataType.Amount
+      && Longs.fromByteArray(add1.data) == Longs.fromByteArray(add2.data))
+      Right(OpcDiff.empty)
+    else
+      Left(GenericError(s"Invalid Assert (eq): DataEntry ${add1.data} is not equal to ${add2.data}"))
+  }
+
+  def isCallerOrigin(context: ExecutionContext)(address: DataEntry): Either[ValidationError, OpcDiff] = {
+    val signer = context.signers.head
+    if (address.dataType != DataType.Address)
+      Left(ContractDataTypeMissMatch)
+    else if (!(address.data sameElements signer.bytes.arr))
+      Left(ContractInvalidCaller)
+    else
+      Right(OpcDiff.empty)
+  }
+
+  def isSignerOrigin(context: ExecutionContext)(address: DataEntry): Either[ValidationError, OpcDiff] = {
+    val signer = context.signers.head
+    if (address.dataType != DataType.Address)
+      Left(ContractDataTypeMissMatch)
+    else if (!(address.data sameElements signer.bytes.arr))
+      Left(ContractInvalidSigner)
+    else
+      Right(OpcDiff.empty)
+  }
+
+  object AssertType extends Enumeration {
+    val GteqZeroAssert = Value(1)
+    val LteqAssert = Value(2)
+    val LtInt64Assert = Value(3)
+    val GtZeroAssert = Value(4)
+    val EqAssert = Value(5)
+    val IsCallerOriginAssert = Value(6)
+    val IsSignerOriginAssert = Value(7)
+  }
+
+  def parseBytes(context: ExecutionContext)
+                (bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, OpcDiff] = bytes.head match {
+    case opcType: Byte if opcType == AssertType.GteqZeroAssert.id && bytes.length == 2
```

Comment:
'bytes.tail.max < data.length && bytes.tail.min >= 0' are repeated in many places.  maybe create a helper function.

Reply:
sure. @Icermli

Reply:
I shall double check and reduce the redundant expression